File: app/src/main/java/com/example/reviewgame/MainPage.java
package com.example.reviewgame;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.example.reviewgame.Classes.CounterTime;

public class MainPage extends AppCompatActivity {
    //Create a laucher to request Permission( lauch a requestion if the permission is or not is granted )
    //In that first line it create a laucher and said that laucher is a permission request
    private final ActivityResultLauncher<String> REQUEST_PERMITION_LAUCHER = registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    //If the permission is granted show a menssage to the user
                    Toast.makeText(this, "Permission granted!", Toast.LENGTH_SHORT).show();
                } else {
                    //ELse show a menssage too
                    Toast.makeText(this, "Permission denied. The counter cannot show notification.", Toast.LENGTH_SHORT).show();
                }
    });


    //Variables

    private boolean isColorChanged = false;

    private ColorStateList defaultButtonColor;

    //END Variables


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.main_page);
        //Calls ask for permition method
        requestNotificationPermission();


        //OBJECTS

        Button start_Review = findViewById(R.id.StartReviewButton);

        //END OBJECTS


        // START BUTTON CODE

        defaultButtonColor = start_Review.getBackgroundTintList();

        // Change button color when click
        // When click in the button
        start_Review.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Intent asks Android, and Android asks CounterTime to execute its routine.
                //And that routine is start the count and update notification bar every second
                Intent serviceCounter = new Intent(MainPage.this, CounterTime.class);
                if(!isColorChanged) {
                    start_Review.setText("Stop Review");
                    int colorWhenClick = Color.parseColor("#FF0000");
                    start_Review.setBackgroundTintList(ColorStateList.valueOf(colorWhenClick));
                    isColorChanged = true;
                    ContextCompat.startForegroundService(MainPage.this, serviceCounter);
                }else{
                    start_Review.setText("Start Review");
                    start_Review.setBackgroundTintList(defaultButtonColor);
                    isColorChanged = false;
                    stopService(serviceCounter);
                }
            }
        });

        //END START BUTTON CODE

    }



    //ASK FOR PERMITION TO USE NOTIFICATION
    private void requestNotificationPermission() {
        //Check if android version needs to accept permission
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            //Check the permission status and see if the result is different of Permission granted ( permissao aceita )
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                //Calls the request_permission_launcher
                REQUEST_PERMITION_LAUCHER.launch(Manifest.permission.POST_NOTIFICATIONS);
            }
        }
    }

}



// REQUEST_PERMITION_LAUCHER Declara o "agente" ou "gerenciador" (`ActivityResultLauncher`) que vai cuidar do processo de pedir uma permissão e receber a resposta.

// No requestNotificationPermission o .lauch e o metodo que de fato dispara ou "lança" a ação, mostrando a caixa de diálogo de permissão para o usuário.

// Passa como argumento do launch a permissão exata que queremos pedir. A classe 'Manifest' contém as constantes com os nomes oficiais das permissões listadas no AndroidManifest.xml.

//O registerFOrActivityResult

// Prepara o lançador com um "contrato" e um "callback".
// O Contrato (new ActivityResultContracts.RequestPermission()): Define que o trabalho a ser feito é pedir uma única permissão.
// O Callback (isGranted -> { ... }): É o bloco de código que será executado com o resultado (true ou false) DEPOIS que o usuário responder ao pedido.

// Contrato no register e o que ele pede o callback e o que faz com a resposta do contrato.

File: app/src/main/java/com/example/reviewgame/MainPage.kt
package com.example.reviewgame

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.example.reviewgame.Classes.CounterTime

class MainPage : AppCompatActivity() {

    // Launcher to request a permission; it will launch a small screen to accept or deny the permission.
    // Create a final (val) variable to make the request.
    // `registerForActivityResult` is for registering an activity result.
    // `ActivityResultContracts` is a class with contracts, and `RequestPermission` is one of them.
    // We send the name of the permission we want to request, and `RequestPermission` will return a boolean.
    // That boolean is put in the 'isGranted' variable.
    private val REQUEST_PERMITION_LAUCHER: ActivityResultLauncher<String> = registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
            // If the permission is granted, show a toast to the user. If not too
            if (isGranted) {
                Toast.makeText(this, "Permissão concedida!", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, "Permissão negada. O app não pode mostrar notificações.", Toast.LENGTH_SHORT).show()
            }
        }



    // Variabless to control the button color
    private var isColorChanged = false
    private var defaultButtonColor: ColorStateList? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_page)

        // Calls the method to request the notification permission
        requestNotificationPermission()

        // Buttons
        val startReviewButton: Button = findViewById(R.id.StartReviewButton)
        val addThemeButton: Button = findViewById(R.id.AddTheme)

        // Tooks the normal button color
        defaultButtonColor = startReviewButton.backgroundTintList

        // When the button startReviewButton is clicked
        startReviewButton.setOnClickListener {
            // Creates the intent to start the service
            val serviceCounter = Intent(this, CounterTime::class.java)

            if (!isColorChanged) {
                startReviewButton.text = "Stop Review"
                val colorWhenClick = Color.parseColor("#FF0000")
                startReviewButton.backgroundTintList = ColorStateList.valueOf(colorWhenClick)
                isColorChanged = true
                ContextCompat.startForegroundService(this, serviceCounter)
            } else {
                startReviewButton.text = "Start Review"
                startReviewButton.backgroundTintList = defaultButtonColor
                isColorChanged = false
                stopService(serviceCounter)
            }
        }


        //When addTheme buttom is clicked
        addThemeButton.setOnClickListener {
            val intent = Intent(this, AddSubject::class.java)
            startActivity(intent)
            //setContentView(R.layout.add_theme)
        }

    }

    // Function to request the notification permission.
    private fun requestNotificationPermission() {
        // If your android version need the request
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // We will see if the permission is granted or not
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                // If yes we send to laucher permission variable the request that we want
                //Manifest.permission.POST_NOTIFICATIONS will be sent to the RequestPermition
                REQUEST_PERMITION_LAUCHER.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
        }
    }
}

//ContextCompart é uma classe que fornece retrocompatibilidade, com ele é como um benjamin para tomada
//Tudo que depende de ter retrocompatibilidade vem dele
//PackageManager é o diretorio oficial de aplicativos
//O permission_granted é um inteiro 0 que no sistema android representa que a permissao foi concebida

